src/Features/CSharp/Portable/Diagnostics/Analyzers/TypeSyntaxSimplifierWalker_Scopes.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.SimplifyTypeNames
+{
+    internal partial class TypeSyntaxSimplifierWalker : CSharpSyntaxWalker
+    {
+        private readonly CSharpSimplifyTypeNamesDiagnosticAnalyzer _analyzer;
+        private readonly SemanticModel _semanticModel;
+        private readonly OptionSet _optionSet;
+        private readonly bool _preferPredefinedTypeInDecl;
+        private readonly bool _preferPredefinedTypeInMemberAccess;
+        private readonly CancellationToken _cancellationToken;
+
+        /// <summary>
+        /// Set of all the type names that this compilation knows about (both through source and
+        /// metadata).  This is used so that when we see <c>Y.Z</c> we can know if <c>Y</c> could be
+        /// a type and thus if we should try to simplify that to <c>X.Z</c> (if <c>X</c> is a base
+        /// class of <c>Y</c>).
+        /// </summary>
+        private readonly HashSet<string> _compilationTypeNames;
```

Comment:
💭 It seems concerning to construct this set for each document. At minimum, we should instead lazily-initialize this once for entire analysis pass (storing the set as part of the `Compilation` analysis context instead of restricting it to a single `SemanticModel` analysis context). We should carefully weigh the benefits of this set against the known cost.

Reply:
that seems reasonable. Note that the computation here is pretty cheap.  the values are already known to the compilation directly from the Decl-Trees.  So the set here is just walking those and putting into our own set.  

i prefer a single set here as it gives us O(1) lookup, wihle `Compilation.TypeNames` is O(N) since it keeps separate collections for each reference.

Reply:
#resolved.